File: src/main/java/typeSpecifications/productAutomatonSpecification/ProductAutomatonNode.java
package typeSpecifications.productAutomatonSpecification;

import typeSpecifications.databaseSpecification.DatabaseNode;
import typeSpecifications.querySpecification.QueryNode;
import typeSpecifications.transducerSpecification.TransducerNode;
import org.javatuples.Triplet;

import java.util.LinkedList;

public class ProductAutomatonNode {

    private Triplet<QueryNode, TransducerNode, DatabaseNode> identifier;
    private boolean initialState;
    private boolean finalState;
    private Double weight;
    private LinkedList<ProductAutomatonEdge> edges;

    public ProductAutomatonNode() {

    }

    /**
     * constructor for a productAutomatonNode. it has the form
     * (queryNode, transducerNode, databaseNode) and has booleans for being an initial or final state
     *
     * @param qNode        the corresponding queryNode
     * @param tNode        the corresponding transducerNode
     * @param dNode        the corresponding databaseNode
     * @param initialState boolean for being an initial state
     * @param finalState   boolean for being a final state
     */
    public ProductAutomatonNode(QueryNode qNode, TransducerNode tNode, DatabaseNode dNode, Boolean initialState, Boolean finalState) {
        identifier = Triplet.with(qNode, tNode, dNode);
        this.initialState = initialState;
        this.finalState = finalState;
        setWeight(null);
        edges = new LinkedList<>();

    }

    /**
     * simple print method.
     */
    public void print() {
        System.out.println(toString());
    }

    public void printWithWeight() {
        System.out.println(toStringWithWeight());
    }

    public String toString() {
        return String.format("(%s, %s, %s)", identifier.getValue0().getIdentifier(), identifier.getValue1().getIdentifier(), identifier.getValue2().getIdentifier());
    }

    public String toStringWithWeight() {
        if (getWeight() == Double.POSITIVE_INFINITY) {
            return String.format("(%s, %s, %s)[INF]", identifier.getValue0().getIdentifier(), identifier.getValue1().getIdentifier(), identifier.getValue2().getIdentifier());
        } else
            return String.format("(%s, %s, %s)[%s]", identifier.getValue0().getIdentifier(), identifier.getValue1().getIdentifier(), identifier.getValue2().getIdentifier(), getWeight());
    }


    public String getIdentifierString() {
        return String.format("%s|%s|%s", identifier.getValue0().getIdentifier(), identifier.getValue1().getIdentifier(), identifier.getValue2().getIdentifier());
    }

    public Triplet<QueryNode, TransducerNode, DatabaseNode> getIdentifier() {
        return identifier;
    }

    public void setWeight(Double weight) {
        this.weight = weight;
    }

    public Double getWeight() {
        return weight;
    }

    public boolean isInitialState() {
        return initialState;
    }

    public boolean isFinalState() {
        return finalState;
    }

    public LinkedList<ProductAutomatonEdge> getEdges() {
        return edges;
    }

    public boolean equals(ProductAutomatonNode otherNode) {
        // compare basic parameters
        if (equalsExcludingEdges(otherNode)) {
            // compare edges
            {
                if (this.edges == null || otherNode.edges == null) {
                    return false;
                }

                if (this.edges.size() != otherNode.edges.size()) {
                    return false;
                }

                return equalsOtherEdgeSet(otherNode.edges);
            }
        }

        return false;
    }

    private boolean equalsOtherEdgeSet(LinkedList<ProductAutomatonEdge> otherEdgeList) {
        int pairsFound = 0;
        for (ProductAutomatonEdge thisEdge : this.edges) {
            for (ProductAutomatonEdge otherEdge : otherEdgeList) {
                if (thisEdge.equals(otherEdge)) {
                    pairsFound++;
                    break;
                }
            }
        }
        return pairsFound == (this.edges.size());
    }

    public boolean equalsExcludingEdges(ProductAutomatonNode otherNode) {
        return this.initialState == otherNode.initialState && this.finalState == otherNode.finalState && this.getWeight().equals(otherNode.getWeight()) && checkIdentifierEquality(otherNode);
    }

    private boolean checkIdentifierEquality(ProductAutomatonNode otherNode) {
        return this.identifier.getValue0().equals(otherNode.identifier.getValue0()) &&
                this.identifier.getValue1().equals(otherNode.identifier.getValue1()) &&
                this.identifier.getValue2().equals(otherNode.identifier.getValue2());
    }

}

File: src/main/kotlin/tin/model/productAutomaton/ProductAutomatonNode.kt
package tin.model.productAutomaton

import tin.model.database.DatabaseNode
import tin.model.query.QueryNode
import tin.model.transducer.TransducerNode
import java.lang.Double.POSITIVE_INFINITY
import java.util.*

/**
 * constructor for a productAutomatonNode. it has the form
 * (queryNode, transducerNode, databaseNode) and has booleans for being an initial or final state
 *
 * @param qNode        the corresponding queryNode
 * @param tNode        the corresponding transducerNode
 * @param dNode        the corresponding databaseNode
 * @param initialState boolean for being an initial state
 * @param finalState   boolean for being a final state
 */

class ProductAutomatonNode(
        queryNode: QueryNode,
        transducerNode: TransducerNode,
        databaseNode: DatabaseNode,
        initialState: Boolean,
        finalState: Boolean,
        ) {

    var identifier: Triple<QueryNode, TransducerNode, DatabaseNode>
    var isInitialState: Boolean = initialState
    var isFinalState: Boolean = finalState
    var weight: Double
    var edges: LinkedList<ProductAutomatonEdge>


    /**
     * simple print method.
     */
    fun print() {
        println(toString())
    }

    fun printWithWeight() {
        println(toStringWithWeight())
    }

    override fun toString(): String {
        return String.format("(%s, %s, %s)", identifier.first.identifier, identifier.second.identifier, identifier.third.identifier)
    }

    fun toStringWithWeight(): String {
        return if (weight == POSITIVE_INFINITY) {
            String.format("(%s, %s, %s)[INF]", identifier.first.identifier, identifier.second.identifier, identifier.third.identifier)
        } else String.format("(%s, %s, %s)[%s]", identifier.first.identifier, identifier.second.identifier, identifier.third.identifier, weight)
    }

    val identifierString: String
        get() = String.format("%s|%s|%s", identifier.first.identifier, identifier.second.identifier, identifier.third.identifier)

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ProductAutomatonNode) return false

        return equalsExcludingEdges(other) && edges.size == other.edges.size && equalsOtherEdgeSet(other.edges)
    }

    private fun equalsOtherEdgeSet(otherEdgeList: LinkedList<ProductAutomatonEdge>?): Boolean {
        var pairsFound = 0
        for (thisEdge in edges) {
            for (otherEdge in otherEdgeList!!) {
                if (thisEdge.equals(otherEdge)) {
                    pairsFound++
                    break
                }
            }
        }
        return pairsFound == edges.size
    }

    fun equalsExcludingEdges(otherNode: ProductAutomatonNode): Boolean {
        return isInitialState == otherNode.isInitialState && isFinalState == otherNode.isFinalState && weight == otherNode.weight && checkIdentifierEquality(otherNode)
    }

    private fun checkIdentifierEquality(otherNode: ProductAutomatonNode): Boolean {
        return identifier.first.equals(otherNode.identifier.first) &&
                identifier.second.equals(otherNode.identifier.second) &&
                identifier.third.equals(otherNode.identifier.third)
    }

    override fun hashCode(): Int {
        var result = identifier.hashCode()
        result = 31 * result + isInitialState.hashCode()
        result = 31 * result + isFinalState.hashCode()
        result = 31 * result + weight.hashCode()
        result = 31 * result + edges.hashCode()
        return result
    }

    init {
        this.identifier = Triple(queryNode, transducerNode, databaseNode)
        this.weight = Double.NaN
        this.edges = LinkedList()
    }
}
